drivers/adc/adc_stm32.c

```diff
@@ -758,71 +778,81 @@ static int start_read(const struct device *dev,
 		return -EINVAL;
 	}
 
-	uint32_t channels = sequence->channels;
-	uint8_t index = find_lsb_set(channels) - 1;
-
-	if (channels > BIT(index)) {
-		LOG_ERR("Only single channel supported");
-		return -ENOTSUP;
-	}
+	data->buffer = sequence->buffer;
+	data->channels = sequence->channels;
+	data->channel_count = 0;
+	data->samples_count = 0;
 
-	adc_stm32_setup_channels(dev, index);
+	for (uint8_t channel_id = 0; channel_id < STM32_CHANNEL_COUNT; channel_id++) {
+		if (data->channels & BIT(channel_id)) {
+			uint32_t channel = __LL_ADC_DECIMAL_NB_TO_CHANNEL(channel_id);
 
-	data->buffer = sequence->buffer;
+			adc_stm32_setup_channel(dev, channel_id);
 
-	uint32_t channel = __LL_ADC_DECIMAL_NB_TO_CHANNEL(index);
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
-	/*
-	 * Each channel in the sequence must be previously enabled in PCSEL.
-	 * This register controls the analog switch integrated in the IO level.
-	 */
-	LL_ADC_SetChannelPreSelection(adc, channel);
+			/*
+			 * Each channel in the sequence must be previously enabled in PCSEL.
+			 * This register controls the analog switch integrated in the IO level.
+			 */
+			LL_ADC_SetChannelPreSelection(adc, channel);
 #elif defined(CONFIG_SOC_SERIES_STM32U5X)
-	/*
-	 * Each channel in the sequence must be previously enabled in PCSEL.
-	 * This register controls the analog switch integrated in the IO level.
-	 * Only for ADC1 instance (ADC4 has no Channel preselection capability).
-	 */
-	if (adc == ADC1) {
-		LL_ADC_SetChannelPreselection(adc, channel);
-	}
+			/*
+			 * Each channel in the sequence must be previously enabled in PCSEL.
+			 * This register controls the analog switch integrated in the IO level.
+			 * Only for ADC1 instance (ADC4 has no Channel preselection capability).
+			 */
+			if (adc == ADC1) {
+				LL_ADC_SetChannelPreselection(adc, channel);
+			}
 #endif
```

Comment:
The ifdeffery here can be decreased by using `IS_ENABLED()` (which I thought was to be preferred anyway).  E.g., 
```c
if (IS_ENABLED(CONFIG_SOC_SERIES_STM32H7X) ||
    (IS_ENABLED(CONFIG_SOC_SERIES_STM32U5X) && adc == ADC1)) {
        LL_ADC_SetChannelPreselection(adc, channel);
}
```
Also less copypasta.

Reply:
This will work only if `LL_ADC_SetChannelPreselection` is defined on all other series.
Code under a `IS_ENABLED(0)` is compiled first and then linked out. If header is not known, it won't compile.

Reply:
Though, I totally agree there are some readability improvements that should be made on this driver.

Reply:
The HAL only provides the functions for specific series?  That's annoying, it defeats the modern style of not using ifdef as much.

A workaround would be to add an empty macro to define `LL_ADC_SetChannelPreselection()`.  It would be nicer if the HAL did that.

Reply:
I agree we need to heavily review this driver to make it more readable. But let's not make current PR more complex

Reply:
Yes, I agree out of scope.  It's not just a case of writing code being added in the PR differently.

Reply:
Sounds good to me too, it needs to be more readable but it's out of scope. Implementing @xyzzy42 suggestion for more efficentl looping will reduce the indent by one tab, which will make it more readable as well.

As a side note, IMO this driver needs an `adc_stm32.h` where the lookup tables and helper functions can move to. This will cleanup this driver quite a bit.